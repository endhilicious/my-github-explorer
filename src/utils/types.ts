export interface User {
  login: string;
  id: number;
  avatar_url: string;
  // Tambahkan properti pengguna lainnya sesuai kebutuhan
}

export interface Repository {
  id: number;
  login: string;
  description: string | null;
  owner: {
    login: string;
    avatar_url: string;
  };
  avatar_url: string;
  stargazers_count: number;
}
