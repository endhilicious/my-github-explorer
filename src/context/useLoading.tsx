import React, { createContext, useContext, useState, ReactNode } from 'react';

import Spinner from '@/common/Spinner';

interface LoadingContextType {
  isLoading: boolean;
  setLoading: (isLoading: boolean) => void;
}

const LoadingContext = createContext<LoadingContextType | undefined>(undefined);

export const useLoading = () => {
  const context = useContext(LoadingContext);
  if (!context) {
    throw new Error('useLoading must be used within a LoadingProvider');
  }
  return context;
};

interface LoadingProviderProps {
  children: ReactNode;
}

export const LoadingProvider: React.FC<LoadingProviderProps> = ({ children }) => {
  const [isLoading, setLoading] = useState(false);

  return (
    <LoadingContext.Provider value={{ isLoading, setLoading }}>
      {isLoading && <Spinner />}
      {children}
    </LoadingContext.Provider>
  );
};
