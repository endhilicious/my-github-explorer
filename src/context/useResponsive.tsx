import React, { createContext, useContext, useEffect, useState } from 'react';

interface ResponsiveContextProps {
  isDesktopView: boolean;
}

const ResponsiveContext = createContext<ResponsiveContextProps | undefined>(undefined);

export const useResponsive = () => {
  const context = useContext(ResponsiveContext);
  if (!context) {
    throw new Error('useResponsive must be used within a ResponsiveProvider');
  }
  return context;
};

export const ResponsiveProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [isDesktopView, setIsDesktopView] = useState<boolean>(true);

  useEffect(() => {
    const checkScreenWidth = () => {
      setIsDesktopView(window.innerWidth > 768);
    };

    checkScreenWidth();

    window.addEventListener('resize', checkScreenWidth);

    return () => {
      window.removeEventListener('resize', checkScreenWidth);
    };
  }, []);

  return (
    <ResponsiveContext.Provider value={{ isDesktopView }}>
      {children}
    </ResponsiveContext.Provider>
  );
};
