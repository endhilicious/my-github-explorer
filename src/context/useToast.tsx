import React, { createContext, useState, useContext, useCallback, ReactNode } from 'react';

import Toast from '@/common/Toast';

interface ToastContextType {
  toastData: {
    message: string;
    visible: boolean;
    type?: string;
  };
  showToast: (message: string, type?: string, time?: number) => void;
  hideToast: () => void;
}

interface ToastProviderProps {
  children: ReactNode;
}

const ToastContext = createContext<ToastContextType | undefined>(undefined);

export const useToast = (): ToastContextType => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error('tolong periksa kembali ToastProvider anda');
  }
  return context;
};

export const ToastProvider: React.FC<ToastProviderProps> = ({ children }) => {
  const [toastData, setToastData] = useState({
    message: '',
    visible: false,
    type: 'success'
  });

  const showToast = useCallback((message: string, type = 'success', time = 3000) => {
    setToastData({ message, visible: true, type });
    setTimeout(hideToast, time);
  }, []);

  const hideToast = useCallback(() => {
    setToastData({ message: '', visible: false, type: 'success' });
  }, []);

  return (
    <ToastContext.Provider value={{ toastData, showToast, hideToast }}>
      {children}
      <Toast />
    </ToastContext.Provider>
  );
};
