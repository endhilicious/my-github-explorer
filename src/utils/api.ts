import axios from 'axios';
import { API_BASE_URL } from './constants';

export const searchUsers = async (query: string, page?: number) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/search/users?q=${query}&per_page=${page || 5}`);
    return response.data.items;
  } catch (error) {
    throw error;
  }
};

export const getUserRepositories = async (username: string) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/users/${username}/repos`);
    console.log('response', response);
    return response.data;
  } catch (error) {
    throw error;
  }
};

export const getUserDetail = async (username: string) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/users/${username}`);
    return response.data;
  } catch (error) {
    throw error;
  }
};

export const getRepositories = async (url: string) => {
  console.log('url',url);
  try {
    const response = await axios.get(url);
    return response.data;
  } catch (error) {
    throw error;
  }
};
