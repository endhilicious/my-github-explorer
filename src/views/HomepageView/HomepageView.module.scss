.home__wrapper {
  max-width: 800px;
  margin: 0 auto;
  padding: 1rem;
  
  main {
    margin-top: 2rem;
  }
}

