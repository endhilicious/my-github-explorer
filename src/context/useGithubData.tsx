import { RepoDetails, UserDetails } from '@/utils/globatTypes';
import React, { createContext, useReducer, useContext, ReactNode } from 'react';

export interface GithubData {
  repos: RepoDetails[];
  selectedUser: UserDetails;
  users: UserDetails[];
}

interface GithubDataAction {
  type: string;
  payload: any;
}

interface GithubDataContextType {
  state: GithubData;
  dispatch: React.Dispatch<GithubDataAction>;
}

const GithubDataContext = createContext<GithubDataContextType | undefined>(undefined);

const initialState: GithubData = {
  repos: [],
  selectedUser: {},
  users: [],
};

const githubDataReducer = (state: GithubData, action: GithubDataAction): GithubData => {
  switch (action.type) {
    case 'ADD_USERS':
      return {
        ...state,
        users: action.payload,
      };
    case 'SELECTED_USER':
      return {
        ...state,
        selectedUser: action.payload,
      };
    case 'ADD_REPOS':
      return {
        ...state,
        repos: action.payload,
      };
    default:
      return state;
  }
};

interface GithubDataProviderProps {
  children: ReactNode;
}

export const useGithubData = (): GithubDataContextType => {
  const context = useContext(GithubDataContext);
  if (!context) {
    throw new Error('useGithubData must be used within a GithubDataProvider');
  }
  return context;
};

export const GithubDataProvider: React.FC<GithubDataProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(githubDataReducer, initialState);

  return (
    <GithubDataContext.Provider value={{ state, dispatch }}>
      {children}
    </GithubDataContext.Provider>
  );
};
