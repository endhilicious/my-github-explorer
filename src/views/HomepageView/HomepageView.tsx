import React from 'react';

import Header from '@/components/Header';
import UserSearch from '@/components/UserSearch';
import UserRepositoryList from '@/components/UserRepositoryList';

import { useLoading } from '@/context/useLoading';
import { useGithubData } from '@/context/useGithubData';
import { useToast } from '@/context/useToast';
import { searchUsers } from '@/utils/api';

import styles from './HomepageView.module.scss';

const Homepage: React.FC = () => {
  const { setLoading } = useLoading();
  const { dispatch } = useGithubData();
  const { showToast } = useToast();

  const handleSearch = async (query: string) => {
    try {
      setLoading(true)
      const newUser = await searchUsers(query);

      dispatch({ type: 'ADD_USERS', payload: newUser });
    } catch (error) {
      console.error('error', error);
      showToast('Error fetching user repositories', 'danger');
    } finally {
      setLoading(false)
    }
  };

  return (
    <div className={styles.home__wrapper}>
      <Header />
      <main>
        <UserSearch onSearch={handleSearch} />
        <UserRepositoryList />
      </main>
    </div>
  );
};

export default Homepage;
