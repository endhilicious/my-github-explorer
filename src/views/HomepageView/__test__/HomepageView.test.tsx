import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';
import HomepageView from '../HomepageView';
import { useGithubData } from '@/context/useGithubData';
import * as apiUtils from '@/utils/api';

jest.mock('../../../context/useLoading', () => ({
  useLoading: jest.fn(() => ({ setLoading: jest.fn() })),
}));

jest.mock('../../../context/useGithubData', () => ({
  useGithubData: jest.fn(() => ({
    state: {
      users: [],
    },
    dispatch: jest.fn()
  })),
}));

jest.mock('../../../context/useToast', () => ({
  useToast: jest.fn(() => ({ showToast: jest.fn() })),
}));

jest.mock('../../../context/useResponsive', () => ({
  useResponsive: jest.fn(() => ({ isDesktopView: true })),
}));

jest.mock('../../../utils/api', () => ({
  searchUsers: jest.fn(),
}));

describe('HomepageView Component', () => {
  it('renders HomepageView component correctly', () => {
    const { container } = render(<HomepageView />);
    expect(container).toMatchSnapshot();
  });

  it('handles search correctly', async () => {
    (useGithubData as any).mockReturnValue({
      state: {
        users: [],
      },
      dispatch: jest.fn(),
    });

    const newUser = {
      id: 1,
      login: 'user123',
      avatar_url: 'https://example.com/avatar.png',
    };
    (apiUtils.searchUsers as any).mockResolvedValue(newUser);

    const { getByTestId } = render(<HomepageView />);
    const searchInput = getByTestId('search-input');
    fireEvent.change(searchInput, { target: { value: 'user123' } });

    fireEvent.click(getByTestId('search-button'));

    await waitFor(() => {
      expect(apiUtils.searchUsers).toHaveBeenCalledWith('user123');
      expect(useGithubData().dispatch).toHaveBeenCalledWith({
        type: 'ADD_USERS',
        payload: newUser,
      });
    });
  });
});
