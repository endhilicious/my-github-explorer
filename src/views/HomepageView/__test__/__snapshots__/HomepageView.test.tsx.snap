// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomepageView Component renders HomepageView component correctly 1`] = `
<div>
  <div
    class="home__wrapper"
  >
    <header
      class="header"
    >
      <h1
        data-testid="header-text"
      >
        Github Repositories Explorer
      </h1>
    </header>
    <main>
      <div
        class="userSearch"
      >
        <div
          class="searchBar"
        >
          <div
            class="inputWrapper"
          >
            <input
              class="searchInput"
              data-testid="search-input"
              placeholder="Search for users..."
              type="text"
              value=""
            />
          </div>
          <button
            class="searchButton"
            data-testid="search-button"
          >
            <span
              class="searchButtonText"
            >
              Search
            </span>
          </button>
        </div>
      </div>
      <div
        class="userRepositoryList"
        data-testid="user-repository-list"
      >
        <h2>
          User Repositories
        </h2>
        <ul />
      </div>
    </main>
  </div>
</div>
`;
